cmd/acr/purge.go

```diff
@@ -264,10 +279,22 @@ func getMatchingRepos(ctx context.Context, repoNames []string, repoRegex string)
 
 // getRepositoryAndTagRegex splits the strings that are in the form <repository>:<regex filter>
 func getRepositoryAndTagRegex(filter string) (string, string, error) {
-	repoAndRegex := strings.Split(filter, ":")
+	// This only selects colons that are not apart of a non-capture group
+	// Note: regexp2 doesn't have .Split support yet, so we just replace the colon with another delimitter \r\n
+	// We choose \r\n since it is an escape sequence that cannot be a part of repo name or a tag
+	noncaptureGroupSupport := regexp2.MustCompile(`(?<!\(\?[imsU-]{0,5}|\[*\^*\[\^*):(?!\]\])`, defaultRegexpOptions)
```

Comment:
This regex looks magic to me. I can understand `imsU` are modifiers but not others. It would be better if there is a documentation on how this regex string is generated.

The `\[*\^*\[\^*)` part seems to address the negate expression as I've noticed in the test case
```
[^[:alpha:]](?ims-U:abc)(?:.*)?:test123[[^:digit:]](?-imUs:.*)
```

In [regexp/syntax](https://pkg.go.dev/regexp/syntax#hdr-Syntax), 

> [[:alpha:]]    ASCII character class
> [[:^alpha:]]   negated ASCII character class

negated class is not `[^[:alpha:]]` but `[[:^alpha:]]`. However, the above regex string can still magically handle it.

Reply:
>  It would be better if there is a documentation on how this regex string is generated.
>

Are you able to see this site? https://regexr.com/6jqp3 This is how I write all of my regex expressions, and there's an explanation tab that explains every component of the expression.

> negated class is not [^[:alpha:]] but [[:^alpha:]]. However, the above regex string can still magically handle it.
>

According to, https://github.com/google/re2/wiki/Syntax, you can also do this: 

<html>
<body>
<!--StartFragment-->
[^[:name:]] | named ASCII class inside negated character class (≡ [:^name:])
-- | --
<!--EndFragment-->
</body>
</html>

So that's where that comes from.  

Basically `\[*\^*\[\^*` means --
```
\[      // match character '[' 
*       // match 0 or more of the preceeding character
\^    // match chaterter '^'
*      // 0 or more of the preceeding character
\[     // match character '['
\^   // match character '^'
*     // 0 or more of the preceeding character
```
So that fragment should match the patterns: 

`[, [[, [^[, [^[^, [[^, ^[`

Now putting that in context with the rest of that group:

`(?<!\(\?[imsU-]{0,5}|\[*\^*\[\^*)`

`(?>!`, this is a negative lookbehind, it means that if it matches *before* the main expression then it is discarded. In this case the main expression is `:`, so that means that any combination of the above pattern plus ':' will get discarded. 

`\(\?[imsU-]{0,5}`, this basically is to handle setting flags in a non-capture group. You can include `-` to "unset", so for example you can write:
```
(?im-sU:re)
```
I won't enumerate all of the patterns that this represents, but basically it's the same idea that I illustrate above. 

Lastly, `(?!\]\]*)` This is a negative lookahead. It means that this group cannot match *after* the main expression, so this is the opposite. It means after you've matched `:`, you can't match with `]` or `]]`, so basically any `:] or :]]` gets discarded.

Hope that explains a bit more. It's not so much magic as it is functional programming, but I guess FP can seem like magic at times 😂